### ISSUE TITLE
System halts on reboot

### ISSUE THREAD

**Issue by sanmiguel** [nerves-project/erlinit#16]
Some pertinent info:

Repo URL: https://github.com/sanmiguel/neuronome/tree/c25c827cf47e2ff4399b9510f90e685a9152bbfd

Using `nerves_system_rpi3: v0.13.0`
Command: `Nerves.Runtime.reboot`

```
iex(2)> Nerves.Runtime.reboot

00:06:25.627 [info]  Elixir.Nerves.Runtime : device told to reboot
:ok
ale: write: Broken pipe
                       iex(3)> erlinit: Ignoring SIGCHLD from pid 0


FATAL ERROR:
erlinit: Unexpected result from sigtimewait: 0 0

CANNOT CONTINUE.

=SUPERVISOR REPORT==== 1-Jan-1970::00:06:25 ===
     Supervisor: {local,'Elixir.SystemRegistry.Supervisor'}
     Context:    shutdown_error
     Reason:     {shutdown,
                     {'Elixir.GenServer',call,
                         [<0.118.0>,
                          {commit,
                              #{'__struct__' =>
                                    'Elixir.SystemRegistry.Transaction',
                                delete_nodes =>
                                    #{'__struct__' => 'Elixir.MapSet',
                                      map =>
                                          #{#{'__struct__' =>
                                                  'Elixir.SystemRegistry.Node',
                                              from => <0.177.0>,
                                              key => collisions,
                                              node =>
                                                  [state,network_interface,
                                                   <<"eth0">>,stats,
                                                   collisions],
                                              parent =>
                                                  [state,network_interface,
                                                   <<"eth0">>,stats]} =>
                                                true,
                                            #{'__struct__' =>
                                                  'Elixir.SystemRegistry.Node',
                                              from => <0.177.0>,key => ifname,
                                              node =>
                                                  [state,network_interface,
                                                   <<"eth0">>,ifname],
                                              parent =>
                                                  [state,network_interface,
                                                   <<"eth0">>]} =>
                                                true,
                                            #{'__struct__' =>
                                                  'Elixir.SystemRegistry.Node',
                                              from => <0.177.0>,key => index,
                                              node =>
                                                  [state,network_interface,
                                                   <<"eth0">>,index],
                                              parent =>
                                                  [state,network_interface,
                                                   <<"eth0">>]} =>
                                                true,
                                            #{'__struct__' =>
                                                  'Elixir.SystemRegistry.Node',
                                              from => <0.177.0>,
                                              key => is_broadcast,
                                              node =>
                                                  [state,network_interface,
                                                   <<"eth0">>,is_broadcast],
                                              parent =>
                                                  [state,network_interface,
                                                   <<"eth0">>]} =>
                                                true,
                                            #{'__struct__' =>
                                                  'Elixir.SystemRegistry.Node',
                                              from => <0.177.0>,
                                              key => is_lower_up,
                                              node =>
                                                  [state,network_interface,
                                                   <<"eth0">>,is_lower_up],
                                              parent =>
                                                  [state,network_interface,
                                                   <<"eth0">>]} =>
                                                true,
                                            #{'__struct__' =>
                                                  'Elixir.SystemRegistry.Node',
                                              from => <0.177.0>,
                                              key => is_multicast,
                                              node =>
                                                  [state,network_interface,
                                                   <<"eth0">>,is_multicast],
                                              parent =>
                                                  [state,network_interface,
                                                   <<"eth0">>]} =>
                                                true,
                                            #{'__struct__' =>
                                                  'Elixir.SystemRegistry.Node',
                                              from => <0.177.0>,
                                              key => is_running,
                                              node =>
                                                  [state,network_interface,
                                                   <<"eth0">>,is_running],
                                              parent =>
                                                  [state,network_interface,
                                                   <<"eth0">>]} =>
                                                true,
                                            #{'__struct__' =>
                                                  'Elixir.SystemRegistry.Node',
                                              from => <0.177.0>,key => is_up,
                                              node =>
                                                  [state,network_interface,
                                                   <<"eth0">>,is_up],
                                              parent =>
                                                  [state,network_interface,
                                                   <<"eth0">>]} =>
                                                true,
                                            #{'__struct__' =>
                                                  'Elixir.SystemRegistry.Node',
                                              from => <0.177.0>,
                                              key => mac_address,
                                              node =>
                                                  [state,network_interface,
                                                   <<"eth0">>,mac_address],
                                              parent =>
                                                  [state,network_interface,
                                                   <<"eth0">>]} =>
                                                true,
                                            #{'__struct__' =>
                                                  'Elixir.SystemRegistry.Node',
                                              from => <0.177.0>,
                                              key => mac_broadcast,
                                              node =>
                                                  [state,network_interface,
                                                   <<"eth0">>,mac_broadcast],
                                              parent =>
                                                  [state,network_interface,
                                                   <<"eth0">>]} =>
                                                true,
                                            #{'__struct__' =>
                                                  'Elixir.SystemRegistry.Node',
                                              from => <0.177.0>,key => mtu,
                                              node =>
                                                  [state,network_interface,
                                                   <<"eth0">>,mtu],
                                              parent =>
                                                  [state,network_interface,
                                                   <<"eth0">>]} =>
                                                true,
                                            #{'__struct__' =>
                                                  'Elixir.SystemRegistry.Node',
                                              from => <0.177.0>,
                                              key => multicast,
                                              node =>
                                                  [state,network_interface,
                                                   <<"eth0">>,stats,multicast],
                                              parent =>
                               [  386.654242] watchdog watchdog0: Watchdog timer stopped
                   [state,network_interface,
        [  386.664212] reboot: System halted
```

**Comment by sanmiguel** [comment]
If the stacktrace from `SystemRegistry` is getting in the way I can trim those lines... I kinda feel like that's a secondary problem caused by the shutdown?

**Comment by fhunleth** [comment]
With any luck I'll be able to reproduce this. If the hang is a side effect from the shutdown error, I'll try to get a nicer error message.

**Comment by fhunleth** [comment]
I couldn't reproduce the error, but I had to stop some genservers from running to get things working enough to test. I inspected the code in `erlinit` and made a few changes to help debug this. I also fixed the behavior for a few unexpected cases to print warnings and carry on rather than giving up. Could you try out this .fw file that I made and capture the output after running Nerves.Runtime.reboot?

https://www.dropbox.com/s/tuwzm694zn4t364/neuronome.fw?dl=0

To use it, just download it, put in a MicroSD card and run `fwup neuronome.fw`. The rest should be familiar.

**Comment by sanmiguel** [comment]
It works! I've truncated the log to just where it starts booting. From there on out it looks like it's absolutely fine. Let me know if you need to see any more of the start log, I will gladly paste it.

```
iex(4)> Nerves.Runtime.reboot

00:00:17.347 [info]  Elixir.Nerves.Runtime [   17.362055] erlinit: sigterm -> reboot
: device told to reboot
[   17.369210] erlinit: waiting 10000 ms for graceful shutdown
:ok
iex(5)> [   22.931257] erlinit: graceful shutdown detected
[   22.938906] erlinit: kill_all
[   22.945010] erlinit: Sending SIGTERM to all processes
[   23.953525] erlinit: Sending SIGKILL to all processes
[   23.961612] erlinit: unmount_all
[   23.968034] erlinit: unmounting /dev/mmcblk0p3(/root)...
[   24.012161] erlinit: unmounting /run(tmpfs)...
[   24.068334] erlinit: unmounting /tmp(tmpfs)...
[   24.098382] erlinit: unmounting /dev/pts(devpts)...
[   24.128343] erlinit: unmounting /sys(sysfs)...
[   24.183384] erlinit: unmounting /proc(proc)...
[   24.209987] watchdog watchdog0: Watchdog timer stopped
[   24.218826] reboot: Restarting system
[    0.000000] Booting Linux on physical CPU 0x0
[    0.000000] Initializing cgroup subsys cpuset
[    0.000000] Initializing cgroup subsys cpu
[    0.000000] Initializing cgroup subsys cpuacct
[    0.000000] Linux version 4.4.50-v7 (fhunleth@halfmarathon) (gcc version 5.3.0 (crosstool-NG v0.10.0) ) #1 SMP PREEMPT Fri Jul 7 22:39:15 EDT 2017
... etc
```

**Comment by sanmiguel** [comment]
Perhaps I should try a completely fresh clone/build and see if that helps? Maybe I've got some wacky stale artefacts somewhere...?

**Comment by sanmiguel** [comment]
OK my local nerves install was `0.3.0`, but upgrading to `0.5.1` and doing a fresh clone/build/burn hasn't made any difference. Same problem as before...

**Comment by fhunleth** [comment]
Thanks for testing this out. The firmware bundle that I sent you had a unreleased version of `erlinit` that would require you to make a custom modification to `nerves_system_br` and `nerves_system_rpi3`. If you're not setup to build this, it can take a long time to replicate. I'll work on getting a new version of those packages out so that this fix can propagate to you.

Also, I realized last night that I told you incorrectly on the option to enable to reboot on failure. It's `--reboot-on-fatal`. I have changed this to be the default since it should be rare enough that no one will know to set it before going to production. I would be curious if adding `--reboot-on-fatal` helps you while you wait for the `erlinit` fix to propagate. I didn't test it.
